ckb-transaction-cobuild/src/lib.rs

```diff
@@ -69,23 +69,18 @@ pub fn fetch_witness_layout() -> Result<WitnessLayout, Error> {
 /// This function can also check the count of SighashAll is one.
 ///
 pub fn fetch_sighash_all() -> Result<SighashAll, Error> {
-    let mut result = None;
+    let mut iter = QueryIter::new(load_witness, Source::Input).filter_map(|witness| {
+        WitnessLayoutReader::from_slice(&witness)
+            .ok()
+            .and_then(|r| match r.to_enum() {
+                WitnessLayoutUnionReader::SighashAll(s) => Some(s.to_entity()),
+                _ => None,
+            })
+    });
 
-    for witness in QueryIter::new(load_witness, Source::Input) {
-        if let Ok(r) = WitnessLayoutReader::from_slice(&witness) {
-            if let WitnessLayoutUnionReader::SighashAll(s) = r.to_enum() {
-                if result.is_some() {
-                    return Err(Error::WrongSighashAll);
-                } else {
-                    result = Some(s.to_entity());
-                }
-            }
-        }
-    }
-    if result.is_some() {
-        return Ok(result.unwrap());
-    } else {
-        return Err(Error::WrongSighashAll);
+    match (iter.next(), iter.next()) {
+        (Some(sighash_with_action), None) => Ok(sighash_with_action),
+        _ => Err(Error::WrongSighashAll),
```

Comment:
These code is difficult to understand. The others looks good to me.

Reply:
It will be better to collect iterator into a vector and make sure the vector has only one element. Otherwise, it fails.

Reply:
This fn uses a QueryIter to iterate over witnesses which is same as the old implementation. The only difference is that it uses a filter_map to extract `SighashAll` variant instead of a temporary var to store the result.

The last match statement checks the iterator by calling iter.next() twice. It expects to find exactly one SighashAll and returns it as the result if successful. If there's more than one or none at all, it returns an error of type Error::WrongSighashAll.

This is a typical pattern for checking the iterator to make sure it has only one element, may I know why you think it's not idiomatic?

Reply:
> It will be better to collect iterator into a vector and make sure the vector has only one element. Otherwise, it fails.

To check and retrieve a fixed number of elements from an iterator, we usually avoid using collect in rust because of the additional allocation. 

Searching some of the projects I starred, I can find many uses of match iter.next(), iter.next() like this:

https://github.com/lapce/lapce/blob/f7d2f4ba863e3a00c9bcb0f3ac1896fd446e7d4c/lapce-core/src/char_buffer.rs#L508

https://github.com/typst/typst/blob/b1256283da047fd4e2c0ba24d981759c745661bb/crates/typst/src/layout/axes.rs#L290